### DISCUSSION TITLE
Potential memory leak: Teleported elements nested inside two `x-ifs` are not properly removed

### DISCUSSION THREAD

**Discussion by mgschoen** [alpinejs/alpine#4287] [7. Bugs (archived)]
Elements that are `x-teleport`-ed but also nested inside an `x-if` should get removed from the DOM, once the condition for the `x-if` becomes falsy. If the following conditions are met however they remain in the DOM:

1. `x-teleport` is nested inside exactly two `x-if`s
2. there exists at least one element between the outer and the inner `x-if`
3. ~~the inner `x-if` depends on values that are defined inside the outer `x-if`~~ (edit: @ekwoka clarified that this doesn't matter)

Minimal example:

```html
<template x-if="showOuter">
    <div x-data>
        <template x-if="true">
            <template x-teleport="#some-other-element">
                <!-- will not be removed from `#some-other-element` when `showOuter` becomes false -->
                <div>Teleported Item</div>
            </template>
        </template>
    </div>
</template>
```

This leads to the teleported element being duplicated every time `x-if` becomes truthy again.

I've reproduced the issue here: https://codepen.io/mgschoen/pen/zYVOZbm

**Comment by ekwoka** [comment]
You need to have intermediary elements between template tags.

Does this continue with that being the case?

edit: [verified](https://awesomealpine.com/play/?coreplugins=0&html=PGRpdiB4LWRhdGE9IntzaG93b3V0ZXI6ZmFsc2UsIHNob3dpbm5lcjpmYWxzZX0iPgogIDxidXR0b24gQGNsaWNrPSJzaG93b3V0ZXJePXRydWUiPnRvZ2dsZSBvdXRlcjwvYnV0dG9uPgogIDxidXR0b24gQGNsaWNrPSJzaG93aW5uZXJePXRydWUiPnRvZ2dsZSBpbm5lcjwvYnV0dG9uPgogIDx0ZW1wbGF0ZSB4LWlmPSJzaG93b3V0ZXIiPgogICAgPGRpdiB4LWRhdGE-CiAgICAgIDxzcGFuPm91dGVyPC9zcGFuPgogICAgICA8dGVtcGxhdGUgeC1pZj0ic2hvd2lubmVyIj4KICAgICAgICA8ZGl2PgogICAgICAgICAgPHRlbXBsYXRlIHgtdGVsZXBvcnQ9ImJvZHkiPgogICAgICAgICAgICA8ZGl2PgogICAgICAgICAgICAgIEhlbGxvIDxzcGFuIHgtdGV4dD0ic2hvd291dGVyIj48L3NwYW4KICAgICAgICAgICAgICA-PHNwYW4geC10ZXh0PSJzaG93aW5uZXIiPjwvc3Bhbj4KICAgICAgICAgICAgPC9kaXY-CiAgICAgICAgICA8L3RlbXBsYXRlPgogICAgICAgIDwvZGl2PgogICAgICA8L3RlbXBsYXRlPgogICAgPC9kaXY-CiAgPC90ZW1wbGF0ZT4KPC9kaXY-Cg&tw=1&script=Y29uc29sZS5sb2coJ2hpJyk7Cg)

To add some clarify, if the inner if is true, and the outer is made false, the ifs cleanup, but not the template.

The x-data between the two x-ifs also does not matter

**Reply by mgschoen** [reply]
Thanks for confirming! You're right, it doesn't matter what scope the inner `x-if` depends on. What does matter though is that there is an `x-data` between the two `x-ifs`.

Updated my initial post accordingly.

**Reply by ekwoka** [reply]
In my tests, that did not matter whether it was there or not.

**Reply by mgschoen** [reply]
Ah true, my bad. Still broken without the x-data. It works if you [leave out the intermediate elements entirely](https://awesomealpine.com/play/?coreplugins=0&html=PGRpdiB4LWRhdGE9IntzaG93b3V0ZXI6ZmFsc2UsIHNob3dpbm5lcjpmYWxzZX0iPgogIDxidXR0b24gQGNsaWNrPSJzaG93b3V0ZXJePXRydWUiPnRvZ2dsZSBvdXRlcjwvYnV0dG9uPgogIDxidXR0b24gQGNsaWNrPSJzaG93aW5uZXJePXRydWUiPnRvZ2dsZSBpbm5lcjwvYnV0dG9uPgogIDx0ZW1wbGF0ZSB4LWlmPSJzaG93b3V0ZXIiPgogICAgPHRlbXBsYXRlIHgtaWY9InNob3dpbm5lciI-CiAgICAgIDx0ZW1wbGF0ZSB4LXRlbGVwb3J0PSJib2R5Ij4KICAgICAgICA8ZGl2PgogICAgICAgICAgSGVsbG8gPHNwYW4geC10ZXh0PSJzaG93b3V0ZXIiPjwvc3BhbgogICAgICAgICAgPjxzcGFuIHgtdGV4dD0ic2hvd2lubmVyIj48L3NwYW4-CiAgICAgICAgPC9kaXY-CiAgICAgIDwvdGVtcGxhdGU-CiAgICA8L3RlbXBsYXRlPgogIDwvdGVtcGxhdGU-CjwvZGl2Pgo&tw=1&script=Y29uc29sZS5sb2coJ2hpJyk7Cg) though.

**Reply by ekwoka** [reply]
interesting, that's more of a happy accident. since the templates should have other elements as their children...
